core/src/com/biglybt/core/networkmanager/impl/tcp/TCPTransportHelper.java

```diff
@@ -408,40 +361,7 @@ private int channelWrite(ByteBuffer buf) throws IOException
 
 		long bytes_read = 0;
 
-		if( enable_efficient_io ) {
-			try{
-				bytes_read = channel.read( buffers, array_offset, length );
-			}
-			catch( IOException ioe ) {
-				//a bug only fixed in Tiger (1.5 series):
-				//http://bugs.sun.com/bugdatabase/view_bug.do?bug_id=4854354
-				String msg = ioe.getMessage();
-				if( msg != null && msg.equals( "A non-blocking socket operation could not be completed immediately" ) ) {
-					enable_efficient_io = false;
-					Logger.log(new LogAlert(LogAlert.UNREPEATABLE, LogAlert.AT_WARNING,
-							"WARNING: Multi-buffer socket read failed; switching to single-buffer mode.\n"
-							+ "Upgrade to JRE 1.5 (5.0) series to fix this problem!"));
-				}
-
-				throw ioe;
-			}
-		}else{
-				//single-buffer mode
-
-			for( int i=array_offset; i < (array_offset + length); i++ ) {
-
-				int data_length = buffers[ i ].remaining();
-
-				int read = channel.read( buffers[ i ] );
-
-				bytes_read += read;
-
-				if( read < data_length ) {
-
-					break;
-				}
-			}
-		}
+		bytes_read = channel.read(buffers, array_offset, length);
```

Comment:
Are the read and write strategies symmetrical? Here they didn't end up using `single-buffer mode`.

Reply:
While you perhaps see this as an interesting voyage of archaeological discovery, I see this code as a heavily overgrown battlefield where I lost many comrades. 

There is a massive amount of crap buried which may or may not now be ready for disposal, the danger lies in the costs of removing something that appears to be 'no longer required' which, when long forgotten and released in 3 months time, causes random application hangs resulting in me (not you) having to invest life-span reducing effort into diagnosing. 

That selector-spin issue took literally weeks, maybe even months, to get into a state the behaved reasonably well on all platforms (follow the tendrils of the spin detection through the codebase if you have a spare moment or two). Of course we were using NIO at the start of its life, when the JVM was still full of bugs. Perhaps they are all fixed now, perhaps not... The safe-selector-mode is currently designed to auto-disable itself every startup in the hope that one day it will not be required, however we're not quite there yet!

Don't get me wrong, I love that you're spending time looking into removing redundant code, just please be sure that the code really is 100% dead. 

Regarding single-buffer-mode, I don't know the code but I do know that whatever it is it currently works, and that's good enough for me.